neuromod-process-spinalcord.nf

```diff
@@ -0,0 +1,304 @@
+PDw = pairPDw
+        .multiMap { it -> 
+                    Nii: tuple(it[0],it[1])
+                    Json: tuple(it[0],it[2])
+                  }
+// Follow the same semantic channel organization for the remaining MTsat file pairs. 
+niiMTS.MTw
+   .join(jsonMTS.MTw)
+   .set {pairMTw}
+
+MTw = pairMTw
+        .multiMap { it -> 
+                    Nii: tuple(it[0],it[1]) 
+                    Json: tuple(it[0],it[2])
+                    }
+
+niiMTS.T1w
+   .join(jsonMTS.T1w)
+   .set {pairT1w}
+
+T1w = pairT1w
+        .multiMap { it -> 
+                    Nii:  tuple(it[0],it[1]) 
+                    Json: tuple(it[0],it[2])
+                    }
+
+// At this point we have PDW, MTw and T1w (each containing Nii and Json sub-channels)
+// for all the spinal-cord data. Now we can combine them as required by the process we 
+// will subject them to.
+
+/** >> EXAMPLE 
+Let's say we have a process named alignInputs that needs nifti files only. 
+**/
+
+// This is how we collect inputs for the alignInputs process: 
+PDw.Nii
+    .join(MTw.Nii)
+    .join(T1w.Nii)
+    // ..._for_... convention indicates that this channel is intended as an input to a process. 
+    .set{mtsat_for_alignment}
+
+/** 
+Now mtsat_for_alignment is a tuple that looks like: 
+  [[sub-01_ses-001, sub-01_ses-001_acq-MToff..nii.gz, sub-01_ses-001_acq-MTon..nii.gz, sub-01_ses-001_acq-T1w..nii.gz]]
+
+This is the reason why input declaration of alignInputs process MUST respect the following order:
+
+input:
+    tuple(sid), file(PDw), file(MTw), file(T1w)
+
+
+*/
+/* >>>>>>> DEFINE PROCESS FOR PUBLISHING OUTPUTS <<<<<<<<<
+
+/*  Process publishOutputs is a typical pattern in DSL2 to put final 
+outputs where they belong to (by copying/moving them from the work folder).
+In BIDS case, it is a derivatives folder `derivatives/sub-01/ses-001/anat/...`. 
+
+We define this process in the body of the main workflow file as it has to 
+access certain variables declared during runtime.
+
+TODO:
+    Test if { sid } instead of "${sid}" can help process access variable when 
+    it is declared to the scope. Same for derivativesDir If so, make publishOutputs a module.
+
+*/
+
+process publishOutputs {
+
+    // Infer session-level file organization.
+    exec:
+        out = getSubSesEntity("${sid}")
+
+    // Again, order matters.
+    // "val" is a Nextflow convention to declare a variable to include in the process
+    // The input variables are outputs in the process. Eg: "segGM" is listed spinalcordsegmentation.process.output. Again, order matters.
+    input:
+      tuple val(sid), \
+      file(T2wSeg)
+
+    // This is where the files will be dropped. Mode move indicates that 
+    // the files will be moved (alternatives are copying or symlinking)
+    // We set overwrite true, if there are files sharing the same name with the 
+    // outputs, they'll be overwritten.
+    publishDir "${derivativesDir}/${out.sub}/${out.ses}anat", mode: 'move', overwrite: true
```

Comment:
@agahkarakuzu when running the pipeline, i noticed that the published output were actually symlinks pointing to the work/ directory. In order to have the physical files (not symlinks), i had to change `mode: 'move'` for `mode: 'copy'`. Do you confirm?

Reply:
Yes @jcohenadad IIRC, copy will create binary copies. I will double check the refs to confirm.

Reply:
is there any reason why you created symlinks for the brain pipeline you created? the problem i see is:
- bob runs the pipeline
- bob thinks that the output files are created under derivatives/
- bob packages the whole dataset and shares it with elsa
- elsa opens the dataset and fails to read the derivatives (because they are symlinks pointing to bob's computer)

Reply:
> is there any reason why you created symlinks for the brain pipeline you created?

According to the documentation, publishDir directive's default mode is creting symlinks. I was expecting `move` option to move the file itself, previously I was using `copy`.  I did not want to create symlinks under the derivatives folder intentionally, I will look into this.